File: scripts/local/schema-gen.go
//go:build ignore

package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"

	"github.com/AtomiCloud/diene.go-consumer/lib/appconfig"
)

func main() {
	output := flag.String("out", "schemas/go-consumer.schema.json", "schema output path")
	flag.Parse()

	schema, err := appconfig.Schema()
	if err != nil {
		fmt.Fprintf(os.Stderr, "compose schema: %v\n", err)
		os.Exit(1)
	}
	raw, err := schema.Marshal()
	if err != nil {
		fmt.Fprintf(os.Stderr, "marshal schema: %v\n", err)
		os.Exit(1)
	}
	if err = os.MkdirAll(filepath.Dir(*output), 0o750); err != nil {
		fmt.Fprintf(os.Stderr, "create schema directory: %v\n", err)
		os.Exit(1)
	}
	raw = append(raw, '\n')
	if err = os.WriteFile(*output, raw, 0o600); err != nil {
		fmt.Fprintf(os.Stderr, "write schema: %v\n", err)
		os.Exit(1)
	}
}

File: scripts/local/schema-gen.ts
import { resolve } from 'node:path';
import { generateJsonSchema } from '@atomicloud/diene.e2e/config';
import { z } from 'zod';
import { applicationRegistry } from '../../src/config/schema';

const argumentsSchema = z.object({ out: z.string().min(1) });
const outIndex = Bun.argv.indexOf('--out');
const arguments_ = argumentsSchema.parse({
  out: outIndex >= 0 ? Bun.argv[outIndex + 1] : resolve(import.meta.dir, '../../schemas/bun-consumer.schema.json'),
});
const schema = generateJsonSchema(applicationRegistry, {
  id: 'https://schemas.atomi.cloud/diene/bun-consumer.schema.json',
});
await Bun.write(arguments_.out, `${JSON.stringify(schema, null, 2)}\n`);
